feat: Add guardian support to registration page

bot190 opened the pull request:
Adds a new Guardians table that creates a Many-To-Many relationship with the User table to link parents to children. It also links one-to-one with the User table to relate to a Parent User. This is intended to allow parents to transition into a more complete role if they like. This should make it easier to add parent specific features in the future. 

Currently parents are only shown on a child's profile. There are plans to add more views of this information that will be more helpful in the future.

bot190 committed:
feat: Add guardian support to registration page

Adds a new Guardians table that creates a Many-To-Many relationship with the User table to link parents to children. It also links one-to-one with the User table to relate to a Parent User. This is intended to allow parents to transition into a more complete role if they like. This should make it easier to add parent specific features in the future. 

Currently parents are only shown on a child's profile. There are plans to add more views of this information that will be more helpful in the future.

msoucy commented on signinapp/model.py:
```diff
@@ -102,6 +116,10 @@ def badges(self) -> list[BadgeAward]:
         ' The badges associated with this user '
         return BadgeAward.query.filter_by(user_id=self.id).all()
 
+    @property
```
Does this need to be a `hybrid_property`?

bot190 replied:
My understanding of hybrid_property is it can be used as part of SQL statements. We shouldn't ever be using this value for SQL stuff, so I don't think it should be a hybrid_property.  Though I don't know if it would be a problem if it was.

msoucy replied:
Let's just say that my understanding was flawed and it suddenly clicked for me. All good!

msoucy commented on signinapp/model.py:
```diff
@@ -144,6 +162,10 @@ def human_readable(self) -> str:
 
+    @classmethod
+    def get_visible_users(cls) -> List[Users]:
+        return cls.query.join(cls.role).filter(Role.visible == True)
```
```py
# Can maybe even use `filter_by` here
return db.session.query.join(cls, Role).filter(Role.visible == True)
```

msoucy commented on signinapp/model.py:
```diff
@@ -164,13 +188,57 @@ def make(name: str, password: str, role: Role, preferred_name: str = None,
 
+    @staticmethod
+    def make_guardian(student: User, name: str, phone_number: str, email: str, preference: int = 1):
+        guardian_limted = Role.from_name("guardian_limited")
```
`limted`?
You could also just call it `guardian_role`

bot190 replied:
The idea is we would add another role for guardians that have actually registered, so this one is "limited". Could definitely use a better name though.

msoucy replied:
I was talking about your variable name, not the role name

msoucy commented on signinapp/model.py:
```diff
@@ -164,13 +188,57 @@ def make(name: str, password: str, role: Role, preferred_name: str = None,
+                            preference=preference, students=[student])
+        db.session.add(guardian)
+
+    def update_guardian(self, student: User):
```
This doesn't update anything about the guardian itself?

bot190 replied:
Shh naming is hard. `add_student`?

msoucy replied:
`add_to_student`? Or a student function `add_guardian`?

msoucy commented on signinapp/model.py:
```diff
@@ -87,6 +97,10 @@ class User(UserMixin, db.Model):
     subteam = relationship("Subteam", back_populates="members")
     awards = relationship("BadgeAward", back_populates="owner")
 
+    guardian = relationship("Guardian", back_populates="user", uselist=False)
```
`guardian_user_data`

msoucy commented on signinapp/model.py:
```diff
@@ -164,13 +188,57 @@ def make(name: str, password: str, role: Role, preferred_name: str = None,
                     approved=approved)
 
+    @staticmethod
+    def make_guardian(student: User, name: str, phone_number: str, email: str, preference: int = 1):
```
It might be nice to have a `Guardian.get_from(name, phone, email)` that gets the guardian if it exists and makes them if not. That could clean up a bunch of code elsewhere.

msoucy commented on signinapp/model.py:
```diff
@@ -164,13 +188,57 @@ def make(name: str, password: str, role: Role, preferred_name: str = None,
+    __tablename__ = "guardians"
+    id = db.Column(db.Integer, primary_key=True)
+    user_id = db.Column(db.Integer, db.ForeignKey("users.id"))
+    preference = db.Column(db.Integer)
```
What is `preference`?

bot190 replied:
Basically order to contact the guardian in, more useful if we allow more than 2 guardians to be added (needs a better UI). 
Would `contact_order` work? Or can you think of a better term?

msoucy replied:
`contact_order` sounds great

bot190 committed:
chore: Cleanup guardian creation and namign

Re-work guardian creation so that the Guardian class drives creating the user and associating it to the Guardian object.

Now creating the guardian object and linked user are done separately from linking the Student user to the guardian so that the registration code is the same whether or not the guardian already existed.

This also renames some code.
